core/trie/src/node_codec.rs

```diff
@@ -92,53 +115,137 @@ impl<H: Hasher> trie_db::NodeCodec<H> for NodeCodec<H> {
 		} else {
-			false
+			partial_enc_it::<N,_>(partial, nb_nibble, NodeKind::BranchNoValue)
+		};
+		let bm_ix = output.len();
+		let mut bm: BM::Buff = Default::default();
+		(0..BM::ENCODED_LEN).for_each(|_|output.push(0));
+		if let Some(value) = maybe_value {
+			value.encode_to(&mut output);
 		};
-		let prefix = branch_node(have_value, children.map(|maybe_child| match maybe_child {
+		BM::encode(children.map(|maybe_child| match maybe_child.borrow() {
 			Some(ChildReference::Hash(h)) => {
 				h.as_ref().encode_to(&mut output);
 				true
 			}
-			Some(ChildReference::Inline(inline_data, len)) => {
-				(&AsRef::<[u8]>::as_ref(&inline_data)[..len]).encode_to(&mut output);
+			&Some(ChildReference::Inline(inline_data, len)) => {
+				inline_data.as_ref()[..len].encode_to(&mut output);
 				true
 			}
 			None => false,
-		}));
-		output[0..3].copy_from_slice(&prefix[..]);
+		}), bm.as_mut());
+		output[bm_ix..bm_ix + BM::ENCODED_LEN].copy_from_slice(&bm.as_ref()[..BM::ENCODED_LEN]);
 		output
 	}
+
+}
+
+// utils
+
+/// Encode and allocate node type header (type and size), and partial value.
+/// It uses an iterator over encoded partial bytes as input.
+fn partial_enc_it<N: NibbleOps, I: Iterator<Item = u8>>(
+	partial: I,
+	nibble_count: usize,
+	node_kind: NodeKind,
+) -> Vec<u8> {
+	let nibble_count = rstd::cmp::min(trie_constants::NIBBLE_SIZE_BOUND, nibble_count);
+
+	let mut output = Vec::with_capacity(3 + (nibble_count / N::NIBBLE_PER_BYTE));
+	match node_kind {
+		NodeKind::Leaf => NodeHeader::Leaf(nibble_count).encode_to(&mut output),
+		NodeKind::BranchWithValue => NodeHeader::Branch(true, nibble_count).encode_to(&mut output),
+		NodeKind::BranchNoValue => NodeHeader::Branch(false, nibble_count).encode_to(&mut output),
+	};
+	output.extend(partial);
+	output
+}
+
+/// Encode and allocate node type header (type and size), and partial value.
+/// Same as `partial_enc_it` but uses non encoded `Partial` as input.
+fn partial_enc<N: NibbleOps>(partial: Partial, node_kind: NodeKind) -> Vec<u8> {
+	let nb_nibble_hpe = (partial.0).0 as usize;
+	let nibble_count = partial.1.len() * N::NIBBLE_PER_BYTE + nb_nibble_hpe;
+
+	let nibble_count = rstd::cmp::min(trie_constants::NIBBLE_SIZE_BOUND, nibble_count);
+
+	let mut output = Vec::with_capacity(3 + partial.1.len());
+	match node_kind {
+		NodeKind::Leaf => NodeHeader::Leaf(nibble_count).encode_to(&mut output),
+		NodeKind::BranchWithValue => NodeHeader::Branch(true, nibble_count).encode_to(&mut output),
+		NodeKind::BranchNoValue => NodeHeader::Branch(false, nibble_count).encode_to(&mut output),
+	};
+	if nb_nibble_hpe > 0 {
+		output.push(N::masked_right(nb_nibble_hpe as u8, (partial.0).1));
+	}
+	output.extend_from_slice(&partial.1[..]);
+	output
+}
+
+/// Radix 16 trie, bitmap encoding implementation,
+/// it contains children mapping information for a branch
+/// (children presence only), it encodes into
+/// a compact bitmap encoding representation.
+pub struct BitMap16(u16);
```

Comment:
`BitMap16` here, yet `ChildBitmap` two lines later. keep your capitalisation consistent! i would suggest sticking with `Bitmap` since it's typically used as a single word.